src/js/background/sync.js

```diff
@@ -0,0 +1,483 @@
+const SYNC_DEBUG = true;
+
+const sync = {
+  storageArea: {
+    area: browser.storage.sync,
+
+    async get(){
+      return await this.area.get();
+    },
+
+    async set(options) {
+      return await this.area.set(options);
+    },
+
+    async getDeletedIdentityList() {
+      const storedArray = await this.getStoredItem("deletedIdentityList");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getIdentities() {
+      const storedArray = await this.getStoredItem("identities");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getDeletedSiteList() { 
+      const storedArray = await this.getStoredItem("deletedSiteList");
+      return (storedArray) ?  storedArray : [];
+    },
+
+    async getCookieStoreIDMap() {
+      const storedArray = await this.getStoredItem("cookieStoreIDmap");
+      return (storedArray) ?  storedArray : {};
+    },
+
+    async getAssignedSites() {
+      const storedArray = await this.getStoredItem("assignedSites");
+      return (storedArray) ?  storedArray : {};
+    },
+
+    async getStoredItem(objectKey) {
+      const outputObject = await this.get(objectKey);
+      if (outputObject && outputObject[objectKey]) 
+        return outputObject[objectKey];
+      return false;
+    },
+
+    async hasSyncStorage(){
+      const inSync = await this.get();
+      return !(Object.entries(inSync).length === 0);
+    },
+
+    async backup(options) {
+      // remove listeners to avoid an infinite loop!
+      await sync.checkForListenersMaybeRemove();
+
+      await updateSyncIdentities();
+      await updateCookieStoreIdMap();
+      await updateSyncSiteAssignments();
+      if (options && options.uuid) 
+        await updateDeletedIdentityList(options.uuid);
+      if (options && options.siteStoreKey) 
+        await addToDeletedSitesList(options.siteStoreKey);
+      if (options && options.undelete) 
+        await removeFromDeletedSitesList(options.undelete);
+      // if (SYNC_DEBUG) {
+      //   const storage = await sync.storageArea.get();
+      //   console.log("inSync: ", storage);
+      //   const localStorage = await browser.storage.local.get();
+      //   console.log("inLocal:", localStorage);
+      //   console.log("idents: ", await browser.contextualIdentities.query({}));
+      // }
+      console.log("Backed up!");
+      await sync.checkForListenersMaybeAdd();
+
+      async function updateSyncIdentities() {
+        const identities = await browser.contextualIdentities.query({});
+        await sync.storageArea.set({ identities });
+      }
```

Comment:
The sync storage seems to get populated by directly asking the contextualIdentities API, this might lead to a case where containers are synchronized that didn't get created by MAC, but by other Container Add-ons - which could lead to unwanted and undefined behavior. E.g. my Temporary Containers Add-ons in Automatic Mode creates a lot of containers for its users, and those should never get synced. Would it maybe be possible to only sync containers created by MAC?

Reply:
As I've seen, MAC doesn't care where the container was made. It'll still show up in the menu.  

I'm playing with your addon now, and I see how that could be messy. As soon as the containers are deleted, it will reset Sync (removing them), but it seems like it would be easy to get one stuck in one sync round, and then be syncing a temporary container around. It would also fill up the deleted uuid storage _fast_, which would waste our already pretty small sync storage.

Reply:
Thanks for looking into it! I'm actually not sure how to sanely tackle that and still allow users to create containers outside of MAC (e.g. by using about:preferences#containers) :-/

Reply:
Something that came to mind: Maybe a possible way would be if MAC provided an [API endpoint](https://github.com/mozilla/multi-account-containers/wiki/API) for excluding certain cookieStoreIds from sync?

Reply:
I think just handling cleaning up of data is key here. I think if a container has been deleted it's historically been a very destructive action so it should likely delete all synced data across multiple machines etc.

Reply:
If that works it'd be great, but @kendallcorner mentioned that there's a list of "deleted uuids" in the sync storage - which I guess makes sense since sync is eventually consistent and afaik there's no reliable way to determine whether all user devices synced the latest changes? And even if there were a way to do that, a user device could stay switched off for a long while, and Temporary Containers users tend to accumulate a lot of cookieStoreIds real quick and most likely exceeding the 100KB sync storage limit.

Other escape-hatches for TC might be: a way to completely disable the sync feature in MAC - preferably also with an API endpoint, so it could be done automatically.

Reply:
I would suggest we give each MAC install it's own UUID then store that in the sync data as the creating install id rather than having a UUID for each container.
This would radically reduce storage cost also.

So a containers unique ID would become `${creatingInstallId}_${userContextId}` where `creatingInstallId` would be an array lookup of UUID's. So: `MacUUIDs: ['uuid1', 'uuid2']` and then indexes 1 and 2 would be used in the `creatingInstallId` rather than the much larger UUID.

That way we can also store in sync indexes of who has had the data synced and mark if it's needing deletion, once all Id's are removed from the 'seen' list we can delete it entirely from the syncing data.

Reply:
That's a neat idea. I think that could work indeed.

Reply:
On install of MAC add uuid for install
Container and SA has been synced with install uuid

Reply:
@jonathanKingston and I discussed, and I've added tracking of identities and site assignments on each install, so when something has been removed (or never existed) on every install, it is removed from the deleted list. This should work for this case.